pkg/model/names.go

```diff
@@ -125,14 +125,14 @@ func (b *KopsModelContext) NameForDNSZone() string {
 }
 
 // IAMName determines the name of the IAM Role and Instance Profile to use for the InstanceGroup
-func (b *KopsModelContext) IAMName(role kops.InstanceGroupRole) string {
+func (b *KopsModelContext) IAMName(igName string, role kops.InstanceGroupRole) string {
 	switch role {
 	case kops.InstanceGroupRoleMaster:
-		return "masters." + b.ClusterName()
+		return igName + "." + b.ClusterName()
 	case kops.InstanceGroupRoleBastion:
-		return "bastions." + b.ClusterName()
+		return igName + "." + b.ClusterName()
 	case kops.InstanceGroupRoleNode:
-		return "nodes." + b.ClusterName()
+		return igName + "." + b.ClusterName()
```

Comment:
Can we test how existing clusters will react to this name change? 

I'm specially worried about Terraform, if I remember correctly the name change would cause the old resources to be garbage collected by Terraform. However this would most likely fail as there's still hosts using the old IAM roles, pretty much causing Terraform based clusters to be stuck in a place where they can't be updated.

Reply:
I will try to test it tomorrow, i didn't have the time do it today.

It's true that it could cause troubles  with Terraform based clusters. Luckily with Terraform we can always manipulate the state before upgrading.

Thanks for raising the issue, i will bring it when discussing this in the office hours.

Reply:
I think the `direct` target would also orphan the old IAM roles with existing clusters - the new roles will be created but the old roles will never be deleted, which might be preferred (I don't recall the behavior of deleting a role still attached to an instance)

Either way, changing the role name means any "downstream" IAM roles assumed by these roles might need their assume role policies updated.

I'm definitely in favor of supporting IG-level IAM roles but we'll have to think through the consequences of a change like this. Perhaps we can make it opt-in via an InstanceGroupSpec field? Otherwise fallback to the old roles and role names.

Reply:
Opt-in for existing clusters and opt-out for new clusters makes sense to me. 

We should also tag our roles so that kops can better reconcile and delete them later.

Reply:
Can we keep the per role roles ? It would allow for a smooth transition for Terraform based clusters until the cluster is rolled out.

> Either way, changing the role name means any "downstream" IAM roles assumed by these roles might need their assume role policies updated.

True, this definitely needs to be well documented.

> Perhaps we can make it opt-in via an InstanceGroupSpec field? Otherwise fallback to the old roles and role names.

If we keep the old roles and role names, we could use the old roles when no additional policy is specified, thus making the `AdditionalPolicy` and `ExternalPolicies` act as an opt-in flag ?
I mean, if a user starts using `AdditionalPolicy` and/or `ExternalPolicies` in an instance group, he probably expects that the role will be different.

Reply:
I changed the implementation, it now creates both per role instance profiles AND per instance group instance profiles.

Then, when linking the launch configuration/launch template with the instance profile, if there's no policy specified for the given instance group it uses the profile associated with the instance group role. If the instance group has additional policy specified it chooses the instance group profile.

This should solve both the terraform based clusters issue and the issue with "downstream" IAM roles assuming.

Except that it will create additional resources (one instance profile/role per instance group), as long as one doesn't specify instance group policy, it should behave exactly the same as before.

cc @rifelpet @rdrgmnzs